vm/devices/hyperv_ic/src/shutdown.rs

```diff
@@ -137,40 +155,53 @@ impl ShutdownChannel {
             .await
             .unwrap();
             match event {
+                Event::StateMachine(Ok(Some(rpc))) => {
+                    let ChannelState::Ready { ref mut state, .. } = &mut self.state else {
+                        panic!("Shutdown request received while not in ready state");
+                    };
+                    if self.pending_shutdown.is_some() {
+                        panic!(
+                            "Shutdown request received while processing previous shutdown request"
+                        );
+                    }
+                    self.pending_shutdown = Some(rpc.1);
+                    *state = ReadyState::SendShutdown(rpc.0);
+                }
                 Event::StateMachine(r) => {
                     r?;
                 }
-                Event::Request(req) => match req {
-                    ShutdownRpc::WaitReady(rpc) => match self.state {
-                        ChannelState::SendVersion | ChannelState::WaitVersion => {
-                            ic.wait_ready.push(rpc)
-                        }
-                        ChannelState::Ready { .. } => rpc.complete(()),
-                    },
-                    ShutdownRpc::Shutdown(rpc) => match self.state {
-                        ChannelState::SendVersion | ChannelState::WaitVersion => {
-                            rpc.complete(ShutdownResult::NotReady)
-                        }
-                        ChannelState::Ready { ref mut state, .. } => match state {
-                            ReadyState::Ready => {
-                                self.pending_shutdown = Some(rpc.1);
-                                *state = ReadyState::SendShutdown(rpc.0);
-                            }
-                            ReadyState::SendShutdown { .. } | ReadyState::WaitShutdown => {
-                                rpc.complete(ShutdownResult::AlreadyInProgress)
-                            }
-                        },
-                    },
+                Event::Request(ShutdownRpc::WaitReady(rpc)) => match &mut self.state {
+                    ChannelState::SendVersion | ChannelState::WaitVersion => {
+                        ic.wait_ready.push(rpc)
+                    }
+                    ChannelState::Ready {
+                        ref mut clients,
+                        shutdown_notification,
+                        ..
+                    } => {
+                        let send_shutdown_notification = &shutdown_notification.0;
+                        let (send, recv) = mesh::oneshot();
+                        clients.push(send);
```

Comment:
Probably should garbage collect the list of clients here so that it doesn't grow without bound:

```
clients.retain(|client| !client.is_closed());
```

Reply:
(Ideally we'd have some collection that would do this for you automatically. But we don't, yet.)